example/src/Api.hs

```diff
@@ -1,34 +1,105 @@
-module Api (API, api, service) where
+module Api where
 
 import ClassyPrelude
-import Composite.Record (Record)
+import Composite.Record (Record, (:->)(Val), pattern (:*:), getVal)
 import Control.Arrow (returnA)
 import Control.Lens (_Unwrapping, each, toListOf, view)
+import Control.Monad.Except (throwError)
 import Control.Monad.Logger (logInfo)
 import Data.Proxy (Proxy(Proxy))
 import Data.Vinyl.Lens (rsubset)
-import Foundation (AppStackM, appMetrics, withDb, fUserRequests)
-import Opaleye ((./=), constant, desc, orderBy, queryTable, restrict, runQuery)
-import Servant (ServerT, Get, JSON, (:>))
-import Types (ApiUserJson(ApiUserJson), DbUser, UserType(UserTypeRegular), cLogin, cUserType, userTable)
+import Foundation
+  ( AppStackM, appMetrics, withDb
+  , fUserCreateRequests, fUserRetrieveRequests, fUserUpdateRequests
+  , fUserDeleteRequests, fUserEnumerateRequests )
+import Opaleye
+  ( (.==), (.&&), constant, desc, limit, orderBy, queryTable, restrict
+  , runDelete, runInsertMany, runQuery, runUpdate )
+import Servant (Capture, Delete, Get, JSON, Post, Put, QueryParam, ReqBody, (:>), (:<|>))
+import Servant.Server (err404)
+import Types
+  ( ApiUser, ApiUserJson(ApiUserJson), DbUser, FId, FLogin, FUserType, WriteColumns
+  , cId, cLogin, cUserType, userTable )
 import qualified System.Metrics.Counter as Counter
 
-type API = "users" :> Get '[JSON] [ApiUserJson]
+type API = "users" :> ( ReqBody '[JSON] ApiUserJson :> Post '[JSON] ()
+                        :<|> Capture "userKey" FId :> Get '[JSON] ApiUserJson
+                        :<|> Capture "userKey" FId :> ReqBody '[JSON] ApiUserJson :> Put '[JSON] ()
+                        :<|> Capture "userKey" FId :> Delete '[JSON] ()
+                        :<|> QueryParam "login" FLogin :> QueryParam "type" FUserType :> Get '[JSON] [ApiUserJson]
+                      )
 
 api :: Proxy API
 api = Proxy
 
-service :: ServerT API AppStackM
-service = do
-  $logInfo "received request for users"
-  liftIO . Counter.inc =<< asks (view fUserRequests . appMetrics) -- Increment the user requests ekg counter
+-- |Convert an id-less User to something that may have a key if it's provided on input; needed
+-- because @Opaleye.Table.TableProperties@ uses writer columns for inserts and deletes
+toWrite :: Maybe FId -> Record ApiUser -> Record WriteColumns
+toWrite userKeyMay user =
+  let convert = constant :: Record ("id" :-> Maybe Int64 ': ApiUser) -> Record WriteColumns
+  in case userKeyMay of
```

Comment:
Suggestion: replace this case statement with `convert $ sequence userKeyMay :*: user`